internal/vfs/osfs/sync_darwin.go

```diff
@@ -0,0 +1,63 @@
+package osfs
+
+import (
+	"golang.org/x/sys/unix"
+
+	"github.com/elastic/go-txfile/internal/vfs"
+)
+
+type syncState struct {
+	noFullSync bool
+}
+
+// Sync uses fnctl or fsync in order to flush the file buffers to disk.
+// According to the darwin fsync man page[1], usage of sync is not safe. On
+// darwin, fsync will only flush the OS file cache to disk, but this won't
+// enforce a cache flush on the drive itself. Without forcing the cache flush,
+// writes can still be out of order or get lost on power failure.
+// According to the man page[1] fcntl with F_FULLFSYNC[2] is required. F_FULLFSYNC
+// might not be supported for the current file system. In this case we will
+// fallback to fsync.
+//
+// [1]: https://www.unix.com/man-page/osx/2/fsync
+// [2]: https://www.unix.com/man-page/osx/2/fcntl
+func (f *File) Sync(flags vfs.SyncFlag) error {
+	if f.state.sync.noFullSync {
+		return f.syncWithFSync()
+	}
+	return f.syncWithFcntl()
+}
+
+func (f *File) syncWithFcntl() error {
+	for {
+		_, err := unix.FcntlInt(f.File.Fd(), unix.F_FULLFSYNC, 0)
+		err = normalizeSysError(err)
+		if err == nil {
+			return nil
+		}
+
+		switch err {
+		// try again
+		case unix.EINTR:
+		case unix.EAGAIN:
+
+		// fallback to fsync?
+		// XXX: always fallback to fsync in future calls?
```

Comment:
This is a good question, from what I've read in the code we will fallback to future calls to fsync.


```

     [EINVAL]		Cmd is F_DUPFD and arg is negative or greater than the maximum allowable
			number (see getdtablesize(2)).

			The argument cmd is F_GETLK, F_SETLK, or F_SETLKW and the data to which
			arg points is not valid, or fildes refers to a file that does not support
			locking.

			The argument cmd is F_PREALLOCATE and the fst_posmode is not a valid
			mode, or when F_PEOFPOSMODE is set and fst_offset is a non-zero value, or
			when F_VOLPOSMODE is set and fst_offset is a negative or zero value.

			The argument cmd is either F_READBOOTSTRAP or F_WRITEBOOTSTRAP and the
			operation was attempted on a non-HFS disk type.
``` 

I am a bit worried here that we switch to fsync, If I understand correctly this could lead to data loss and we wont have any traces or log notifying the user that we have moved to a lesser fsync strategy, from the fnctl manpage this look like a permanent situation.

Should we report that?
Should it be part of check at the beginning?

Reply:
It's kinda tricky. According to different sources F_FULLFSYNC only works if disk is internal + supported by the FS layer (e.g. NFS doesn't seem to support the F_FULLFSYNC). Some libs (e.g. SQLite) just use F_FULLFSYNC always and fallback to fsync on failure only.

Any use of `fsync` is indeed subject to data loss here.

Things we can do here:
- always use F_FULLFSYNC + fsync on failure
- when introducing stats/user logging, report some 'sync' status for OS X, warning about not using F_FULLFSYNC

Reply:
I think I would go the SQLite route here which is F_FULLSYNC + fsync on failure, I am OK with the performance hit that could occur, but I think it's an expected last resort. Also introducing stats and logging will help diagnose the issue.

I think I am scared of _not knowing_ which mode I am currently in if we do a durable failover.

Reply:
👍 

The question is: how durable is the failover? Like, what happens to the error status of the written pages if F_FULLFSYNC did fail due to an earlier IO error? Will the kernel reset the pages IO error state on return, such that fsync is guaranteed to succeed?

As FreeBSD is known to not reset the error status on failed writes (+1 for FreeBSD) + darwin uses some of the FreeBSD code base,  I guess it's somewhat safe to assume we get the correct/wanted behavior on failed writes.

Anyways, I did change the code to always do F_FULLFSYNC first and fallback to fsync depending on the error code. Checking other projects code bases, there is no common way on how F_FULLFSYNC+fscyn fallback should be done for real.

Reply:
@urso 👍  I believe we will get a better idea of the behavior in the wild, so your work on stats/metrics for this feature is really important.